Make resampler infer the input sampling period and buffer length

leandro-lucarella-frequenz opened the pull request:
The input sampling period is calculated by counting the total received samples, and dividing the total resampling time until the internal buffer is full by the total received samples.

To store all the source properties (sampling period, sampling start, total received samples) a new class is used (SourceProperties) and this information is passed to the resampling function, so it can have more information about the source to perform a proper resampling.

This calculation is performed once and then remains constant, but at some point we could add a timer to re-calculate it.

This also improves slightly the documentation and validation of the ResamplingConfig class.

The add_timeseries() methods now also takes a name as a way to identify different sources in log messages (the actor uses the channel name as timeseries name).

We also update the resampling buffer length based on the input sampling period, resampling period, and max data age in periods, so it can hold all the past data without wasting any space.

Fixes #55.

llucax committed:
Add a maximum resampling buffer size

This is mainly because in a near future the resampling buffer size will
dynamically change based on the input sampling rate, so we need to make
sure it doesn't get out of hand in the event of some unlikely high input
sampling rate combined with a very low frequency resampling.

We also add a warning size, for which we just emit a warning in the logs
this is to also try to minimize the presence of very big buffers without
affecting the running apps (unless again, it gets out of hand). When
this happens users should either try to use a different resampling
period to keep buffer smaller, or change the (default) warning buffer
length if they know it is still manageable.

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

llucax committed:
Pass the whole ResamplerConfig to the ResamplingFunction

Instead of only passing the resampling period, we pass the whole
ResamplerConfig so the resampling function has more information on how
the resampling is being done to make better decisions.

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

llucax committed:
Calculate the input sampling period

The input sampling period is calculated by counting the total received
samples, and dividing the total resampling time until the internal
buffer is full by the total received samples.

To store all the source properties (sampling period, sampling start,
total received samples) a new class is used (SourceProperties) and this
information is passed to the resampling function, so it can have more
information about the source to perform a proper resampling.

This calculation is performed once and then remains constant, but at
some point we could add a timer to re-calculate it.

This also improves slightly the documentation and validation of the
ResamplingConfig class.

The add_timeseries() methods now also takes a name as a way to identify
different sources in log messages (the actor uses the channel name as
timeseries name).

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

llucax committed:
Update the buffer length based on the input sampling period

When the input sampling period is know, we can adjust the internal
buffer to have a good size to be able to store all the requested
resampling and not waste space if the initial buffer was too big.

If we are upsampling, one sample could be enough for back-filling, but
we store max_data_age_in_periods using the input sampling period, so
the resampling functions can do more complex inter/extrapolation if they
need to.

If we are upsampling, we want a buffer that can hold
max_data_age_in_periods * resampling_period_s seconds of data, and we
have one sample every input_sampling_period_s, so we use a buffer length
of:
max_data_age_in_periods * resampling_period_s / input_sampling_period_s

The buffer size, like the input sampling period, is only calculated
once. The buffer data is copied to a new buffer when/if the buffer needs
resizing. The new size is validated against config.warn_buffer_len and
config.max_buffer_len and will emit a warning if it goes above any of
those and if it goes above config.max_buffer_len it will be truncated to
config.max_buffer_len as a safety measure.

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/_resampling.py:
```diff
@@ -334,6 +480,94 @@ def add_sample(self, sample: Sample) -> None:
+            or props.received_samples
+            < config.resampling_period_s * config.max_data_age_in_periods
+            or len(self._buffer) < self._buffer.maxlen
+        ):
```
should this also check `props.received_samples` is equal to zero? or is it redundant?

daniel-zullo-frequenz replied:
it is redundant as `max_data_age_in_periods` is ensured to be greater than 1 and `resampling_period_s` is checked for positive

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/_resampling.py:
```diff
@@ -334,6 +480,94 @@ def add_sample(self, sample: Sample) -> None:
+        ):
+            return False
+
+        samples_time_delta = now - props.sampling_start
```
I wonder if `now > props.sampling_start` should be checked/asserted to avoid a zero or negative delta.

leandro-lucarella-frequenz replied:
Yeah, I could add a check in case there is a race between the first sample is received and the timer to do the resampling fires :+1:

leandro-lucarella-frequenz replied:
Fixed.

ela-kotulska-frequenz commented on src/frequenz/sdk/timeseries/_resampling.py:
```diff
@@ -388,6 +480,42 @@ def add_sample(self, sample: Sample) -> None:
         self._buffer.append(sample)
+        if self._source_properties.sampling_start is None:
+            self._source_properties.sampling_start = sample.timestamp
+        self._source_properties.received_samples += 1
```
What about:
```
if props.sampling_period_s is None:
   self._source_properties.received_samples += 1
```

leandro-lucarella-frequenz replied:
I was thinking that maybe we want to re-calculate the input sampling rate in the future again, maybe regularly at some defined interval, and maybe we want to take into account the whole history (probably not) or just the last T time. For now I think it makes sense to just keep adding, because the variable counts all the samples received so far, it might be confusing that at some point it stops counting, but in the future we might want to stop or probably even better reset it to adjust the input sampling rate from now and then to make sure we are not sticking to a wrong period forever (if we were unlucky enough to happen to measure it when there was some networking issue or something like that).

ela-kotulska-frequenz commented on src/frequenz/sdk/timeseries/_resampling.py:
```diff
@@ -388,6 +480,42 @@ def add_sample(self, sample: Sample) -> None:
+            return False
+
+        samples_time_delta = now - props.sampling_start
+        props.sampling_period_s = (
+            samples_time_delta.total_seconds()
+        ) / props.received_samples
+
+        return True
```
Could you print info log with `props`? To see in logs what is computed `sampling_period_s`

leandro-lucarella-frequenz replied:
Yeah, that's a good idea.

leandro-lucarella-frequenz replied:
Fixed.

ela-kotulska-frequenz commented on src/frequenz/sdk/timeseries/_resampling.py:
```diff
@@ -517,6 +517,58 @@ def _update_source_sample_period(self, now: datetime) -> bool:
+            )
+
+        new_buffer_len = max(1, math.ceil(new_buffer_len))
+        if new_buffer_len > config.max_buffer_len:
+            _logger.warning(
+                "The new buffer length (%s) for timeseries %s is too big, using %s instead",
+                new_buffer_len,
+                self._name,
+                config.max_buffer_len,
+            )
```
Shouldn't it be an error?
And the exception should be raised.

leandro-lucarella-frequenz replied:
I don't think an exception should be raised because we don't get anything from that, there isn't much to do by the caller if this happens (except maybe rebuild the whole thing increasing the `max_buffer_len` or adjusting the `resampling_period_s`, but those are more deployment things IMHO), but I could turn it into an error. It is not really fatal, but it is true that it can degrade the quality of the resampling if it happens so maybe it is better to shout louder.

Does this look acceptable?

leandro-lucarella-frequenz replied:
Oh, I see you found a bug, my intention was to truncate the buffer to `max_buffer_len`, which I now see it's not being done! I will fix that too, and should probably add a test too.

leandro-lucarella-frequenz replied:
Fixed (without raising an exception for now)

leandro-lucarella-frequenz commented:
Something else I forgot to mention in the PR is to check if the `DEFAULT_XXX_BUFFER_LEN` make any sense, I assumed the typical input sampling period is about 0.2s and the typical resampling period is 1s, so with a window of 3, a length of 16 should be enough (default initial length), so 8 times more than that should be suspicious (warning length) and 64 times that looks very excessive, like something is really not right (error length).

@ela-kotulska-frequenz @sahas-subramanian-frequenz @idlir-shkurti-frequenz @tiyash-basu-frequenz maybe you can double check these numbers.

llucax committed:
Fix wrong logger creation

Loggers need to be created using logging.getLogger() instead of
logging.Logger(), otherwise they don't go through the configuration
system, and they can't be controlled in tests for example.

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

leandro-lucarella-frequenz commented:
I added a new commit to also fix a log of wrongly created loggers, that were getting me crazy as there was no way to see/capture the logs in the tests.

llucax committed:
Update release notes

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

tiyash-basu-frequenz commented:
>I assumed the typical input sampling period is about 0.2s and the typical resampling period is 1s, so with a window of 3, a length of 16 should be enough (default initial length), so 8 times more than that should be suspicious (warning length) and 64 times that looks very excessive, like something is really not right (error length).

Assuming 16 is the length of the buffer holding the 0.2s res samples, if we want a resampled window of 3x1 second, then it sounds alright. Also, 8 times and 64 times as warning and error thresholds sounds decent to me.

llucax committed:
Add a hack to ignore logs unrelated to the tests

Some resampling tests started getting some log errors about some gRPC
task not being properly closed, which are clearly artefacts from other
tests that shouldn't appear there.

For example:

        assert config.initial_buffer_len == init_len
>       assert caplog.record_tuples == [
            (
                "frequenz.sdk.timeseries._resampling",
                logging.WARNING,
                f"initial_buffer_len ({init_len}) is bigger than "
                f"warn_buffer_len ({DEFAULT_BUFFER_LEN_WARN})",
            )
        ]
E       assert [('asyncio', 40, "Task was destroyed but it is pending!\ntask: <Task pending name='raw-component-data-8' coro=<MicrogridGrpcClient._component_data_task() done, defined at /home/luca/devel/sdk/src/frequenz/sdk/microgrid/client/_client.py:298> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7f9224472ca0>()]>>"), ('asyncio', 40, "Task was destroyed but it is pending!\ntask: <Task pending name='raw-component-data-18' coro=<MicrogridGrpcClient._component_data_task() done, defined at /home/luca/devel/sdk/src/frequenz/sdk/microgrid/client/_client.py:298> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7f9224472c10>()]>>"), ('frequenz.sdk.timeseries._resampling', 30, 'initial_buffer_len (257) is bigger than warn_buffer_len (128)')] == [('frequenz.sdk.timeseries._resampling', 30, 'initial_buffer_len (257) is bigger than warn_buffer_len (128)')]
E         At index 0 diff: ('asyncio', 40, "Task was destroyed but it is pending!\ntask: <Task pending name='raw-component-data-8' coro=<MicrogridGrpcClient._component_data_task() done, defined at /home/luca/devel/sdk/src/frequenz/sdk/microgrid/client/_client.py:298> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7f9224472ca0>()]>>") != ('frequenz.sdk.timeseries._resampling', 30, 'initial_buffer_len (257) is bigger than warn_buffer_len (128)')
E         Left contains 2 more items, first extra item: ('asyncio', 40, "Task was destroyed but it is pending!\ntask: <Task pending name='raw-component-data-18' coro=<Microgr...k/microgrid/client/_client.py:298> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7f9224472c10>()]>>")
E         Full diff:
E           [
E         +  ('asyncio',
E         +   40,
E         +   'Task was destroyed but it is pending!\n'
E         +   "task: <Task pending name='raw-component-data-8' "
E         +   'coro=<MicrogridGrpcClient._component_data_task() done, defined at '
E         +   '/home/luca/devel/sdk/src/frequenz/sdk/microgrid/client/_client.py:298> '
E         +   'wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at '
E         +   '0x7f9224472ca0>()]>>'),
E         +  ('asyncio',
E         +   40,
E         +   'Task was destroyed but it is pending!\n'
E         +   "task: <Task pending name='raw-component-data-18' "
E         +   'coro=<MicrogridGrpcClient._component_data_task() done, defined at '
E         +   '/home/luca/devel/sdk/src/frequenz/sdk/microgrid/client/_client.py:298> '
E         +   'wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at '
E         +   '0x7f9224472c10>()]>>'),
E            ('frequenz.sdk.timeseries._resampling',
E             30,
E             'initial_buffer_len (257) is bigger than warn_buffer_len (128)'),
E           ]

tests/timeseries/test_resampling.py:113: AssertionError

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

leandro-lucarella-frequenz commented:
Rebased and added a hack to the tests to cope with some spurious errors in logs from unrelated tests. Also enabled auto-merge.

leandro-lucarella-frequenz commented:
Unrelated, but FYI, I noticed now the tests are way slower than before. Not all of them, so my guess is, since I think I say some wait time to gather battery status in recent PRs, that maybe tests are now subject to this extra waiting time.

If this is the case we really need to start using `time-machine` and `async-solipsism` in all tests that depends on time where time can be faked via these 2 libraries. You can look at the resampling tests for an example.
